### PR TITLE
Add test to check public APIs against master

### LINKED ISSUE #58617: [rfc] Bug developers about release notes for new APIs
Right now it's very easy to add (or remove!) a public API to `torch` without anyone noticing. #58610 has a proof-of-concept script that walks all the reachable APIs in `torch` and prints them out. We could keep a log of these on master in S3 then diff that against PRs. This would then give us a listing of new APIs per PR. We could then take it a step further and error out unless the commit message contains something like:

```
### Release Notes
* <insert text here>
```

This would help on gathering release notes for releases and prevent devs from accidentally adding public APIs

### PR DESCRIPTION
**Pull request by driazati**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60927**
* #61088

This implements/fixes #58617. If a PR adds a new bindings to the Python APIs, this will show up as a failing test, which they can fix by adding release notes to the PR body. The check errs on the side of caution in the face of flakiness, so if it can't reach S3 or doesn't have the env variables it expects it just quits.

Also this fixes an issue in the `torch._C` api bindings test where it wasn't actually testing the list defined in the file.

Differential Revision: [D29497017](https://our.internmc.facebook.com/intern/diff/D29497017)

### CODE REVIEW

File: scripts/list_apis.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+        attrs = sorted(list(set(attrs)))
+        add_item(path, name, out)
+
+        for attr in attrs:
+            try:
+                next_obj = getattr(obj, attr)
+                crawl_helper(next_obj, attr, path + [name], seen_objects, out, relevant_module)
+            except ModuleNotFoundError as e:
+                err_print(e)
```
Comment (albanD) [inline scripts/list_apis.py:111]: Are these error useful? Or just triggering when we try to access attributes that we should not?

File: scripts/list_apis.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                err_print(f"ERROR: unaccessible attribute {get_name(path, name)}")
+                out.errors.append(
+                    CrawlError(
+                        reason="unaccessible attribute", path=get_name(path, name)
+                    )
+                )
+
+    elif inspect.isclass(obj):
+        attrs = dir(obj)
```
Comment (albanD) [inline scripts/list_apis.py:128]: You don't add the class itself here?

File: scripts/list_apis.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    except Exception as e:
+        err_print(e)
+        pass
+
+    return attrs
+
+
+def crawl_helper(
+    obj: Any,
+    name: str,
```
Comment (albanD) [inline scripts/list_apis.py:86]: Is name always the same as path[-1] here ?

File: scripts/list_apis.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+from types import ModuleType
+from collections import namedtuple
+from typing import List, Any, Set
+
+
+CrawlError = namedtuple("CrawlError", ("reason", "path"))
+CrawlResult = namedtuple("CrawlResult", ("public", "private", "errors", "module"))
+
+
+def err_print(*args: Any) -> None:
```
Comment (albanD) [inline scripts/list_apis.py:15]: There are a lot of error printing happening here. Do we want to actually keep that? Or should we silence these? Or have a list of expected errors?

**Review by albanD — COMMENTED** [review]
Thanks for working on this! Very interesting read!

File: scripts/list_apis.py (hunk 2 above)
Comment (driazati) [reply scripts/list_apis.py:128]: it should be (like with modules), good catch

File: scripts/list_apis.py (hunk 3 above)
Comment (driazati) [reply scripts/list_apis.py:86]: `name` is the next attribute to look at in the `path` (which is just used for reporting back the fully qualified name), so on the next recursive call the previous `name` will be part of `path`

**Comment by driazati** [comment]
@driazati has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29497017).

File: scripts/list_apis.py (hunk 1 above)
Comment (driazati) [reply scripts/list_apis.py]: mostly the errors come from objects / modules bound from C that Python can't deal with (e.g. missing attributes such as `__file__` on C modules) and some that explode if you try to access their attributes in any way (e.g. the TorchBind namespace). These are all generally private anyways so we don't really care

**Comment by driazati** [comment]
@driazati has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29497017).

**Review by walterddr — COMMENTED** [review]
I might not be the best to review. but did we consider reusing some logic from target_determinator, which uses modulefinder to crawl the entire codebase backwards from test files. I haven't tried, but i think it could be adjust to crawl from `torch`?

File: .github/templates/linux_ci_workflow.yml.j2 [hunk 5]
```diff
@@ -493,3 +495,80 @@ jobs:
...
+          bash /tmp/ecr-login.sh
+          rm /tmp/ecr-login.sh
+      - name: Chown workspace
+        run: |
+          # Ensure the working directory gets chowned back to the current user
+          docker run --rm -v "$(pwd)":/v -w /v "${ALPINE_IMAGE}" chown -R "$(id -u):$(id -g)" .
+      - name: Checkout PyTorch
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0 # deep clone, to allow sharding to use git rev-list
```
Comment (samestep) [inline .github/templates/linux_ci_workflow.yml.j2:521]: this job doesn't do anything with sharding, right? could you either remove this line or modify this comment to clarify why it's necessary?

File: .github/templates/linux_ci_workflow.yml.j2 [hunk 6]
```diff
@@ -493,3 +495,80 @@ jobs:
...
+            "${DOCKER_IMAGE}" \
+            bash -c "sudo chown -R jenkins . && pip install dist/*.whl && python scripts/list_apis.py --module torch --public | sort > apis.log"
+      - name: Chown workspace
+        run: |
+          # Ensure the working directory gets chowned back to the current user
+          docker run --rm -v "$(pwd)":/v -w /v "${ALPINE_IMAGE}" chown -R "$(id -u):$(id -g)" .
+      - name: Archive APIs
+        run: |
+          zip apis.zip apis.log
+      - uses: driazati/upload-artifact-s3@50adbe4ef0b6d9221df25c18c5fc528dfcb7c3f8
```
Comment (samestep) [inline .github/templates/linux_ci_workflow.yml.j2:559]: it looks like there are three different versions of `upload-artifact-s3` that we use:

- `driazati/upload-artifact-s3@50adbe4ef0b6d9221df25c18c5fc528dfcb7c3f8` (this one), in one place
- `driazati/upload-artifact-s3@21c31d0a7bcb056ca50bd6ce197ba6507c26a1be`, in [one place](https://github.com/pytorch/pytorch/blob/4a544df00dee565e6ccb73e971ff734346b16474/.github/templates/linux_ci_workflow.yml.j2#L484)
- `seemethere/upload-artifact-s3@9d7ceb0ab39c2c88d93ef7792b27425b27d59162`, in [two](https://github.com/pytorch/pytorch/blob/4a544df00dee565e6ccb73e971ff734346b16474/.github/templates/linux_ci_workflow.yml.j2#L188) [places](https://github.com/pytorch/pytorch/blob/4a544df00dee565e6ccb73e971ff734346b16474/.github/templates/windows_ci_workflow.yml.j2#L85)

If it is necessary to use all these different versions, could you add code comment(s) explaining why? And if not, would it be a good idea to consolidate to use a single version?

**Comment by samestep** [comment]
why is `list_apis.py` in `scripts/`? that causes it to not be checked by either Flake8 or `mypy`

File: scripts/list_apis.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    parser = argparse.ArgumentParser(
+        description="Recursively list all reachable objects in a Python module (library)."
+        + " This is used in PyTorch releases to determine the API surface changes"
+    )
+    parser.add_argument("--module", help="module to crawl", required=True)
+
+    def add_flag(name: str, default: bool, help: str) -> None:
+        parser.add_argument(f"--{name}", dest=name, help=help, action="store_true")
+        parser.add_argument(f"--no-{name}", dest=name, help=help, action="store_false")
+        parser.set_defaults(**{name: default})
```
Comment (samestep) [inline scripts/list_apis.py:260]: I don't quite understand this: in `main` above, you use this for `public`, `private`, and `errors`, but in `main` above, you implicitly coerce all three of those to booleans before using them, so the `--no-` variant seems to not do anything; could you clarify why you add it?

File: scripts/list_apis.py [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+    def get_name(self, path: List[str], name: str) -> str:
+        names = [item.name for item in path]
+        return ".".join(names + [name])
+
+    def should_skip(self, obj: Any, path: List[Item]) -> bool:
+        """
+        Return whether 'obj' is relevant to 'relevant_module'. If 'obj' is not a
+        module, then get its module from the __module__ attribute. Then check if
+        the module is part of the relevant_module by comparing their __file__ paths.
```
Comment (samestep) [inline scripts/list_apis.py:189]: what is `relevant_module`?

File: scripts/list_apis.py [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+
+        if id(obj) in self.apis:
+            # Some things are listed twice, usually as a side-effect of importing
+            # it directly while still in the library. This uses a heuristic of
+            # shorter-path-is-better to pick the best one
+            prev_len = len(self.apis[id(obj)])
+            curr_len = len(path) + 1
+            if curr_len >= prev_len:
+                # It's a longer path and this already has an entry, so don't
+                # both with it any further
```
Comment (samestep) [inline scripts/list_apis.py:158]: typo I think
```suggestion
                # bother with it any further
```

File: scripts/list_apis.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+            try:
+                module = importlib.import_module(module_name)
+            except ModuleNotFoundError as e:
+                # Likely we hit an incorrectly named module defined in C (so
+                # the name given to the CPython API isn't actually importable)
+                return True
+
+
+        if module == self.module:
+            # If the object is a direct child of the concered module, then don't
```
Comment (samestep) [inline scripts/list_apis.py:222]: typo I think
```suggestion
            # If the object is a direct child of the concerned module, then don't
```

File: scripts/list_apis.py [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+
+                for attr in attrs:
+                    to_add.append(path + [Crawler.Item(name=attr, obj=None)])
+
+        for index, path in enumerate(to_add):
+            self.apis[f"attr-{index}"] = path
+
+    def get_submodules(self, module: ModuleType) -> List[ModuleType]:
+        """
+        Some modules aren't imported directly into the parent (i.e. torch.fx). So
```
Comment (samestep) [inline scripts/list_apis.py:81]:
```suggestion
        Some modules aren't imported directly into the parent (e.g. torch.fx). So
```

File: scripts/list_apis.py [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+        attributes.
+        """
+        to_add = []
+        for path in self.apis.values():
+            item = path[-1]
+            obj = item.obj
+            if inspect.isclass(obj):
+                # We don't want to re-add class attributes (these don't have an
+                # associated id() to de-duplicate them with, so we have to do it
+                # manually
```
Comment (samestep) [inline scripts/list_apis.py:70]:
```suggestion
                # manually)
```

File: scripts/list_apis.py (hunk 7 above)
Comment (samestep) [reply scripts/list_apis.py:260]: oh never mind, sorry; I didn't see that you set the `default` for these when you call `add_flag` below

File: test/test_public_bindings.py [hunk 13]
```diff
@@ -1,10 +1,81 @@
...
+
+        difference = master_public_bindings.symmetric_difference(current_public_bindings)
+        msg = textwrap.dedent(f"""
+            New public API(s) were added: {', '.join(difference)}
+
+            To silence this error, add a release notes section to your PR body:
+
+                ### Release Notes
+                * the new feature""")
+        self.assertTrue(difference != set() and "### Release Notes" not in PR_BODY, msg)
```
Comment (samestep) [inline test/test_public_bindings.py:76]: I'm a bit confused about this; shouldn't it instead look like the following?
```suggestion
        self.assertTrue(difference == set() or "### Release Notes" in PR_BODY, msg)
```

**Comment by driazati** [comment]
> modulefinder

`modulefinder` kind of does the same thing (though it has a bug where it crashes on the `mpl_toolkits` module), but it misses modules unreachable from the source like `torch.fx`

```python

File: scripts/list_apis.py (hunk 8 above)
Comment (driazati) [reply scripts/list_apis.py:189]: forgot to update the comment after OOP-ing everything, it's the main module that's being tracked (so this should be `relevant_module` -> `self.module`)

File: .github/templates/linux_ci_workflow.yml.j2 (hunk 6 above)
Comment (driazati) [reply .github/templates/linux_ci_workflow.yml.j2:577]: We should definitely consolidate: #61384

**Comment by driazati** [comment]
closing since this seems like a recipe for flakiness

**Pull request closed without merging**